src/_arraykit.c

```diff
@@ -491,6 +491,318 @@ isna_element(PyObject *Py_UNUSED(m), PyObject *arg)
+        new_indexers = np.empty(len(array), dtype=np.int64)                  # 3
+
+        num_found = 0                                                        # 4
+
+        for i, element in enumerate(array):                                  # 5
+            if element_locations[element] == num_unique:                     # 6
+                element_locations[element] = num_found                       # 7
+                num_found += 1                                               # 8
+
+            if num_found == num_unique:                                      # 9
+                return positions, array                                      # 10
+
+            new_indexers[i] = element_locations[element]                     # 11
+
+        x = get_index_screen(element_locations, num_unique)                  # 12
+        return x, new_indexers                                               # 13
+    */
+    PyArrayObject *array;
+    PyArrayObject *positions;
+
+    static char *kwlist[] = {"array", "positions", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O!O!:get_new_indexers_and_screen", kwlist,
+                &PyArray_Type, &array,
+                &PyArray_Type, &positions
+        ))
+    {
+        return NULL;
+    }
+
+    npy_intp num_unique = PyArray_SIZE(positions);
+
+    if (num_unique > PyArray_SIZE(array))
+    {
+        // This algorithm is only optimal if the number of unique elements is
+        // less than the number of elements in the array.
+        // Otherwise, the most optimal code is ``np.unique(array, return_index=True)[1]``
+        // and we don't want to re-implement that in C.
+        PyErr_SetString(
+                PyExc_ValueError,
+                "Number of unique elements must be less than or equal to the length of the array"
+                );
+        return NULL;
+    }
+
+    if (PyArray_TYPE(array) != NPY_INT64)
+    {
+        PyErr_SetString(
+                PyExc_ValueError,
+                "Array must be of type np.int64"
+                );
+        return NULL;
+    }
+
+    npy_intp dims = {num_unique};
+    PyArrayObject *element_locations = (PyArrayObject*)PyArray_Empty(
+            1,                                // ndim
+            &dims,                            // shape
+            PyArray_DescrFromType(NPY_INT64), // dtype
+            0                                 // fortran
+            );
+    if (element_locations == NULL) {
+        return NULL;
+    }
+
+    PyObject *num_unique_pyint = PyLong_FromLong(num_unique);
+    if (num_unique_pyint == NULL) {
+        Py_DECREF(element_locations);
+        return NULL;
+    }
+
+    // We use ``num_unique`` to signal that we haven't found the element yet
+    // This works, because each element must be 0 < num_unique.
+    if (PyArray_FillWithScalar(element_locations, num_unique_pyint)) {
+        Py_DECREF(element_locations);
+        Py_DECREF(num_unique_pyint);
+        return NULL;
+    }
```

Comment:
I think all of this can be done without using `PyLong_FromLong`, and instead building a C array of size equal to `element_locations` * bytes per int64, and then just passing that array as `data` to PyArray_SimpleNewFromData. Something like this:

npy_int64 data[num_unique] = { num_unique };
PyArrayObject *element_locations = PyArray_SimpleNewFromData(1, npy_intp_dims, NPY_INT64, data)

Reply:
This array will need to be returned to caller in some cases, so I am concerned about where this data lives. In your example, the data is on the stack, and thus will be de-allocated once the function is exited, leading to segfaults.

If I create the array using `malloc`, then `delete` will never be called because I don't know when the array will be garbage collected.

Reply:
In general, I'm not sure the cost of creating one additional PyObject is worth the extra code complexity

Reply:
Here is a questionably better way to fill that array that lets NumPy manage the buffer's memory correctly. At least it is a low fewer lines of code!

```c
PyObject *array = PyArray_Zeros(1, dims, dtype, 0)
npy_int64 *array_buffer = (npy_int64*)PyArray_DATA((PyArrayObject*)array);
npy_int64 *end = array_buffer + count;
while (array_buffer < end) {*array_buffer++ = num_unique}
```